rtp_media_server: adding module

jchavanton opened the pull request:
<!-- Kamailio Pull Request Template -->



#### Pre-Submission Checklist
- [x] Commit message has the format required by CONTRIBUTING guide
- [x] Commits are split per component (core, individual modules, libs, utils, ...)
- [x] Each component has a single commit (if not, squash them into one commit)
- [x] No commits to README files for modules (changes must be done to docbook files
in `doc/` subfolder, the README file is autogenerated)

#### Type Of Change
- [x] New module

#### Checklist:

Tested with various sofphones : Ekiga, Linphone, CSipSimple

#### Description
By combining  Kamailio, oRTP and mediastreamer2 
this module is providing some very good foundation to support RTP endpoint and various media processing features.

This  seems like a great idea for various use cases like IoT, IVR or other specific needs.

Kamailio is handling everything related to SIP/SDP (the module is adding some SDP parsing) as well as providing a scripting engine.

oRTP is providing RTP endpoints compliant with (RFC 3550)

MediaStreamer2, even if written in C is quite a high level library, because it is implementing a framework for audio processing using graphs of filters, filters can be to do various things.
Similar to playing with lego blocks :)

* Support for most free and some non free codecs can be added easily.
* Work to bridge calls is already in progress in the module.
* Mobile phone support ARM CPU 
* other embedded scenario should be supported 
* much more ...

Mediastream2 is creating one thread per call "msticker", this can work smoothly in Kamailio even if it is forking processes.
Shared memory allocation is supported using wrapper around malloc/free used by the libraries.

Some extra work that needs to be done shorty
* Syncronization with locks is not completed (minor task, only access to a linked list needs to be syncronized properly)
* Memory leak not tested properly (not take the time to test with valgrind yet)

This project was started last year, I think it is time to submit it, I will surely find the time to do the extra work needed shortly.


Example config using the features already implemented 
```
event_route[rms:start] {
        xnotice("[rms:start] play ...\n");
        rms_play("/tmp/reference_8000.wav", "rms:after_play");
};

event_route[rms:after_play] {
        xnotice("[rms:after_play] play done...\n");
        rms_hangup();
};

route {
        if (t_precheck_trans()) {
                t_check_trans();
                exit;
        }
        t_check_trans();
        if (is_method("INVITE") &amp;&amp; !has_totag()) {
                if (!rms_answer()) {
                        t_reply("503", "server error");
                }
        }

        if (is_method("BYE")){
                rms_media_stop();
        }
}
```

henningw commented:
Thank you Julien for the contribution of this module. I will do a quick first review of the code.

The git master is currently in a feature freeze in preparation for the upcoming 5.2 release, so no new features can be integrated. So I would suggest to finish the mentioned remaining steps in your branch.

Did I understand the purpose of the module correctly, as its implements a RTP endpoint directly in Kamailio? This is kind of a new approach (for Kamailio) I think, the existing RTP handling modules only place the control channel into kamailio, and handle the RTP processing in a separate daemon. 

It would suggest to write an e-mail to sr-dev, describing your module, the approach and getting feedback.

henningw commented on src/modules/rtp_media_server/rms_media.c:
```diff
@@ -0,0 +1,320 @@
+	return 1;
+}
+
+#define MS_UNUSED(x) ((void)(x))
```
What is the purpose of this define?

jchavanton replied:
I modified the includes for mediastreamer2 and included the file that was defining this macro instead.

henningw commented on src/modules/rtp_media_server/rms_media.c:
```diff
@@ -0,0 +1,320 @@
+	ortp_memory_functions.free_fun = ptr_shm_free;
+	ortp_set_memory_functions(&ortp_memory_functions);
+	ortp_init();
+	vars = ortp_malloc(sizeof(shared_global_vars_t));
```
Why you use the ortp_malloc() for allocation of Kamailio module structures?
The malloc can also fail.

jchavanton replied:
no good reason, ortp_malloc is using shm_malloc, I replaced the 2 occurrences with shm_malloc for clarity from Kamailio module context

henningw commented on src/modules/rtp_media_server/rms_media.h:
```diff
@@ -0,0 +1,103 @@
+//#include <mediastreamer2/mediastream.h>
+#include <ortp/ortp.h>
+#include <ortp/port.h>
+#define MS_UNUSED(x) ((void)(x))
```
see above

jchavanton replied:
removed, since we now include mscommon.h

henningw commented on src/modules/rtp_media_server/rms_sdp.c:
```diff
@@ -0,0 +1,301 @@
+//"a=rtpmap:96 opus/48000/2\r\n"
+//"a=fmtp:96 useinbandfec=1\r\n";
+
+static char *rms_shm_strdup(char *source)
```
Can't you use the core shm_str_dup() instead?

jchavanton replied:
The reason is that shm_str_dup does not garanty terminated 'char*', this is one less thing to worry about. The module is not trying to optimize string like Kamailio is when parsing messages.

jchavanton replied:
Maybe we could add another helper function in the core ?

jchavanton replied:
I just did refactoring for now , we can clarify if it is best to use helper function from the core, add new ones or not ...

henningw commented on src/modules/rtp_media_server/rms_sdp.c:
```diff
@@ -0,0 +1,301 @@
+			payload_type_number);
+	body->len += strlen(sdp_m);
+
+	body->s = pkg_malloc(body->len + 1);
```
pkg_malloc can fail

jchavanton replied:
check that none was left unchecked

henningw commented on src/modules/rtp_media_server/rtp_media_server.c:
```diff
@@ -0,0 +1,716 @@
+	return (rtn);
+}
+
+int rms_str_dup(str *dst, str *src, int shared)
```
please use the pkg_str_dup() and shm_str_dup() in the core

jchavanton replied:
Same topic as before, since they do not convert to `\0` terminated `char *` they are not flexible in this case.

henningw commented on src/modules/rtp_media_server/rtp_media_server.c:
```diff
@@ -0,0 +1,716 @@
+	return 1;
+}
+
+str reply_headers = {0, 0};
```
unused

henningw commented on src/modules/rtp_media_server/rtp_media_server.c:
```diff
@@ -0,0 +1,716 @@
+}
+
+str reply_headers = {0, 0};
+str headers = str_init("Max-Forwards: 70" CRLF);
```
This and the following vars are only used in one functions. Please consider moving them to the respective functions, they don't need to be globally visible.

henningw commented on src/modules/rtp_media_server/rms_sdp.c:
```diff
@@ -0,0 +1,301 @@
+	char *copy;
+	if(!source)
+		return NULL;
+	copy = (char *)ortp_malloc(strlen(source) + 1);
```
see other comment

henningw commented:
Hello Julien, thank you for updating the pull request. I also saw your message on sr-dev, so far no reply from the other developers. About the code - I will have another closer look on Friday. if there are no objections by then from the list we can discuss about integrating it into the master branch.

jchavanton commented:
Thanks, this is good news !

So far there was only a brief discussion in slack, but nothing valuable enough to share in the mailing list.
Mainly I shared my optimism about the fact that these libs are also designed to build server side components.
The very minimalist feature set currently included was questioned, I believe that extending the features is not such a big endeavor considering the architecture of the routing script and the libraries. 

I still need to address some comments you have made.

jchavanton commented:
revisited the synchronization and added a few more error handling ...
We are sure we want to remove  the small voice file ? `157K Bach_10s_8000.wav`

SurendraPlivo commented:
Just curious to know that rtp_media_server module will not conflict with
rtpengine module.

On Fri, 9 Nov 2018 at 12:11 PM, Julien Chavanton <notifications@github.com>
wrote:

> revisited the synchronization and added a few more error handling ...
> We are sure we want to remove the small voice file ? 157K
> Bach_10s_8000.wav
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kamailio/kamailio/pull/1701#issuecomment-437265721>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AZoKAh3lBVGahBWPHLQkyWSn-BvSAHR1ks5utSOPgaJpZM4YF7gV>
> .
>
-- 
Surendra Tiwari
VoIP Application Developer,
Plivo Inc., 201 Mission Street, *Suite 230*, San Francisco - 94105, USA
Web: www.plivo.com | Twitter: @plivo

henningw commented:
> We are sure we want to remove the small voice file ? 157K Bach_10s_8000.wav

I would just adding a link in the README to e.g. this, they provide voice files in different languages:
http://www.voiptroubleshooter.com/open_speech/

jchavanton commented:
> Just curious to know that rtp_media_server module will not conflict with rtpengine module.
> On Fri, 9 Nov 2018 at 12:11 PM, Julien Chavanton ***@***.***> wrote: revisited the synchronization and added a few more error handling ... We are sure we want to remove the small voice file ? 157K Bach_10s_8000.wav — You are receiving this because you are subscribed to this thread. Reply to this email directly, view it on GitHub <[#1701 (comment)](https://github.com/kamailio/kamailio/pull/1701#issuecomment-437265721)>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AZoKAh3lBVGahBWPHLQkyWSn-BvSAHR1ks5utSOPgaJpZM4YF7gV> .
> -- Surendra Tiwari VoIP Application Developer, Plivo Inc., 201 Mission Street, *Suite 230*, San Francisco - 94105, USA Web: [www.plivo.com](http://www.plivo.com) | Twitter: @plivo

Hi Surendra, RTP engine module is encoding the SDP and transferring it to RTP engine server application.

RTP Media Server is starting an RTP enpoint in a thread from a Kamailio forked process.

There nothing related that can conflict, expect the fact that they will both play with the SDP, of course you can not use both module on the same call at the same time :) but you could use both module on the same Kamailio instance/config

SurendraPlivo commented:
ohk. Because what I will try with module by converting out of band dtmf to
in band dtmf.

On Fri, 9 Nov 2018 at 12:28 PM, Julien Chavanton <notifications@github.com>
wrote:

> Just curious to know that rtp_media_server module will not conflict with
> rtpengine module.
> On Fri, 9 Nov 2018 at 12:11 PM, Julien Chavanton ***@***.***> wrote:
> revisited the synchronization and added a few more error handling ... We
> are sure we want to remove the small voice file ? 157K Bach_10s_8000.wav —
> You are receiving this because you are subscribed to this thread. Reply to
> this email directly, view it on GitHub <#1701 (comment)
> <https://github.com/kamailio/kamailio/pull/1701#issuecomment-437265721>>,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AZoKAh3lBVGahBWPHLQkyWSn-BvSAHR1ks5utSOPgaJpZM4YF7gV
> .
> -- Surendra Tiwari VoIP Application Developer, Plivo Inc., 201 Mission
> Street,
> <https://maps.google.com/?q=201+Mission+Street,++Suite+230+,+San+Francisco+-+94105,+USA&entry=gmail&source=g>*Suite
> 230
> <https://maps.google.com/?q=201+Mission+Street,++Suite+230+,+San+Francisco+-+94105,+USA&entry=gmail&source=g>*,
> San Francisco - 94105, USA
> <https://maps.google.com/?q=201+Mission+Street,++Suite+230+,+San+Francisco+-+94105,+USA&entry=gmail&source=g>
> Web: www.plivo.com | Twitter: @plivo <https://github.com/plivo>
>
> Hi Surendra, RTP engine module is encoding the SDP and transferring it to
> RTP engine server application.
>
> RTP Media Server is starting an RTP enpoint in a thread from a Kamailio
> forked process.
>
> There nothing related that can conflict, expect the fact that they will
> both play with the SDP, of course you can not use both module on the same
> call at the same time :) but you could use both module on the same Kamailio
> instance/config
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/kamailio/kamailio/pull/1701#issuecomment-437268557>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AZoKAr5Qsrri6VoDmoEGUGTtHT0rQIqJks5utSeGgaJpZM4YF7gV>
> .
>
-- 
Surendra Tiwari
VoIP Application Developer,
Plivo Inc., 201 Mission Street, *Suite 230*, San Francisco - 94105, USA
Web: www.plivo.com | Twitter: @plivo

jchavanton commented:
> ohk. Because what I will try with module by converting out of band dtmf to in band dtmf. On Fri, 9 Nov 2018 at 12:28 PM, Julien Chavanton <notifications@github.com> wrote:
> Just curious to know that rtp_media_server module will not conflict with rtpengine module. On Fri, 9 Nov 2018 at 12:11 PM, Julien Chavanton ***@***.***> wrote: revisited the synchronization and added a few more error handling ... We are sure we want to remove the small voice file ? 157K Bach_10s_8000.wav — You are receiving this because you are subscribed to this thread. Reply to this email directly, view it on GitHub <#1701 (comment) <[#1701 (comment)](https://github.com/kamailio/kamailio/pull/1701#issuecomment-437265721)>>, or mute the thread https://github.com/notifications/unsubscribe-auth/AZoKAh3lBVGahBWPHLQkyWSn-BvSAHR1ks5utSOPgaJpZM4YF7gV . -- Surendra Tiwari VoIP Application Developer, Plivo Inc., 201 Mission Street, <https://maps.google.com/?q=201+Mission+Street,++Suite+230+,+San+Francisco+-+94105,+USA&entry=gmail&source=g>*Suite 230 <https://maps.google.com/?q=201+Mission+Street,++Suite+230+,+San+Francisco+-+94105,+USA&entry=gmail&source=g>*, San Francisco - 94105, USA <https://maps.google.com/?q=201+Mission+Street,++Suite+230+,+San+Francisco+-+94105,+USA&entry=gmail&source=g> Web: [www.plivo.com](http://www.plivo.com) | Twitter: @plivo <https://github.com/plivo> Hi Surendra, RTP engine module is encoding the SDP and transferring it to RTP engine server application. RTP Media Server is starting an RTP enpoint in a thread from a Kamailio forked process. There nothing related that can conflict, expect the fact that they will both play with the SDP, of course you can not use both module on the same call at the same time :) but you could use both module on the same Kamailio instance/config — You are receiving this because you commented. Reply to this email directly, view it on GitHub <[#1701 (comment)](https://github.com/kamailio/kamailio/pull/1701#issuecomment-437268557)>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AZoKAr5Qsrri6VoDmoEGUGTtHT0rQIqJks5utSeGgaJpZM4YF7gV> .
> -- Surendra Tiwari VoIP Application Developer, Plivo Inc., 201 Mission Street, *Suite 230*, San Francisco - 94105, USA Web: [www.plivo.com](http://www.plivo.com) | Twitter: @plivo

Not sure I understand your last comment

miconda commented:
Although I didn't have time to look at the code, I am fine with merging it, being a lot of time till 5.3 release that allows testing and tuning, if someone is interested.

Actually, I wanted to write here to just say that @jchavanton has write access to kamailio already. So he can also merge and commit already.

jchavanton commented:
I believe everything in the review was addressed

I did remove the audio file and added a link to download one.

I added another commit to support multiple actions this will be needed by some action like DTMF capture that could interrupt play and probably many other future commands.

henningw commented:
Thank you!
